sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,376 @@
+        output = {}
+        if partition_info:
+            output['event_hub_path'] = partition_info[b'name'].decode('utf-8')
+            output['id'] = partition_info[b'partition'].decode('utf-8')
+            output['beginning_sequence_number'] = partition_info[b'begin_sequence_number']
+            output['last_enqueued_sequence_number'] = partition_info[b'last_enqueued_sequence_number']
+            output['last_enqueued_offset'] = partition_info[b'last_enqueued_offset'].decode('utf-8')
+            output['last_enqueued_time_utc'] = datetime.datetime.utcfromtimestamp(
+                float(partition_info[b'last_enqueued_time_utc'] / 1000))
+            output['is_empty'] = partition_info[b'is_partition_empty']
+        return output
+
+    def close(self):
+        # type:() -> None
+        self._conn_manager.close_connection()
+
+
+class ConsumerProducerMixin(object):
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def _check_closed(self):
+        if self.closed:
+            raise EventHubError("{} has been closed. Please create a new one to handle event data.".format(self._name))
+
+    def _open(self):
+        """Open the EventHubConsumer/EventHubProducer using the supplied connection.
+
+        """
+        # pylint: disable=protected-access
+        if not self.running:
+            if self._handler:
+                self._handler.close()
+            self._create_handler()
+            self._handler.open(connection=self._client._conn_manager.get_connection(  # pylint: disable=protected-access
+                self._client._address.hostname,
+                self._client._create_auth()
+            ))
+            while not self._handler.client_ready():
+                time.sleep(0.05)
+            self._max_message_size_on_link = self._handler.message_handler._link.peer_max_message_size \
+                                             or constants.MAX_MESSAGE_LENGTH_BYTES  # pylint: disable=protected-access
+            self.running = True
+
+    def _close_handler(self):
+        if self._handler:
+            self._handler.close()  # close the link (sharing connection) or connection (not sharing)
+        self.running = False
+
+    def _close_connection(self):
+        self._close_handler()
+        self._client._conn_manager.reset_connection_if_broken()  # pylint: disable=protected-access
+
+    def _handle_exception(self, exception):
+        if not self.running and isinstance(exception, compat.TimeoutException):
+            exception = errors.AuthenticationException("Authorization timeout.")
+        return _handle_exception(exception, self)
+
+    def _do_retryable_operation(self, operation, timeout=100000, **kwargs):
+        # pylint:disable=protected-access
+        # timeout equals to 0 means no timeout, set the value to be a large number.
+        timeout_time = time.time() + (timeout if timeout else 100000)
```

Comment:
This last test `(timeout if timeout else 100000)` doesn't make sense since you made 100000 the default value of timeout in the method signature.
If 0 is an invalid value, raise it don't hide it under a 100000. Because if that's the customer doing it, he/she definitely didn't want to see 100000. If it's your own code doing it, fix your own code :)

Reply:
0 is not an invalid value - if I understand the comment on line 347 correctly, 0 indicates "no timeout", so the default of a "large value" is used in it's place.... though this doesn't seem correct to me either. Will double check....

Reply:
+ @yunhaoling + @YijunXieMS

Reply:
yes, we should be able to set timeout_time to `None`.
I checked that the timeout_out ops are used in `send`/`open`. The send can handle timeout_time = None and open doesn't use timeout_timeout.

Let's make the default to None, and then
`timeout_time = (time.time() + timeout) if timeout else None`

Reply:
Fixed!